dlib/dnn/loss.h

```diff
@@ -3447,6 +3447,526 @@ namespace dlib
+                        {
+                            loss += -scale * options.lambda_cls * safe_log(out_data[c_idx]);
+                            g[c_idx] = scale * options.lambda_cls * (out_data[c_idx] - 1);
+                        }
+                        else
+                        {
+                            loss += -scale * options.lambda_cls * safe_log(1 - out_data[c_idx]);
+                            g[c_idx] = scale * options.lambda_cls * out_data[c_idx];
+                        }
+                    }
+                }
+            }
+        };
+    }
+
+    template <template <typename> class... TAG_TYPES>
+    class loss_yolo_
+    {
+        static void list_tags(std::ostream& out) { impl::yolo_helper_impl<TAG_TYPES...>::list_tags(out); }
+
+    public:
+
+        typedef std::vector<yolo_rect> training_label_type;
+        typedef std::vector<yolo_rect> output_label_type;
+
+        constexpr static size_t tag_count() { return impl::yolo_helper_impl<TAG_TYPES...>::tag_count(); }
+
+        loss_yolo_() {};
+
+        loss_yolo_(const yolo_options& options) : options(options) { }
+
+        template <
+            typename SUB_TYPE,
+            typename label_iterator
+            >
+        void to_label (
+            const tensor& input_tensor,
+            const SUB_TYPE& sub,
+            label_iterator iter
+        ) const
+        {
+            DLIB_CASSERT(sub.sample_expansion_factor() == 1, sub.sample_expansion_factor());
+            std::vector<yolo_rect> dets_accum;
+            output_label_type final_dets;
+            for (long i = 0; i < input_tensor.num_samples(); ++i)
+            {
+                dets_accum.clear();
+                impl::yolo_helper_impl<TAG_TYPES...>::tensor_to_dets(input_tensor, sub, i, options, dets_accum);
+
+                // Do non-max suppression
+                std::sort(dets_accum.rbegin(), dets_accum.rend());
+                final_dets.clear();
+                for (size_t j = 0; j < dets_accum.size(); ++j)
+                {
+                    if (overlaps_any_box_nms(final_dets, dets_accum[j]))
+                        continue;
+
+                    final_dets.push_back(dets_accum[j]);
+                }
+
+                *iter++ = std::move(final_dets);
+            }
+        }
+
+        template <
+            typename const_label_iterator,
+            typename SUBNET
+        >
+        double compute_loss_value_and_gradient (
+            const tensor& input_tensor,
+            const_label_iterator truth,
+            SUBNET& sub
+        ) const
+        {
+            DLIB_CASSERT(input_tensor.num_samples() != 0);
+            DLIB_CASSERT(sub.sample_expansion_factor() == 1);
+            double loss = 0;
+            for (long i = 0; i < input_tensor.num_samples(); ++i)
+            {
+                impl::yolo_helper_impl<TAG_TYPES...>::tensor_to_loss(input_tensor, truth, sub, i, options, loss);
```

Comment:
After debugging for a while, I found a problem, after this line, I should add:
```c++
++truth;
```
So that the next iteration uses the correct bounding boxes, otherwise it's using the bounding boxes of the first image in the batch for all the mini batch... I hope that is the problem, running some tests now...

Reply:
I think that was the problem, it's learning slowly, but for the first time I'm getting results that look like this:
![image](https://user-images.githubusercontent.com/1671644/122921801-e75ead80-d39d-11eb-9fa1-5317a57572d2.png)
🤞

Reply:
Is this learning from scratch? Or finetuning an existing model?

Reply:
Fine-tuning, the training code it's a bit different from the one published here. I don't know what will be best for an example program, though... I think training from scratch is simpler, and people can refer to [`dnn_introduction3_ex`](http://dlib.net/dnn_introduction3_ex.cpp.html) for how to do fine-tuning.

Reply:
Well it's a good sign if you're training completely from scratch and you're getting results like that.

Reply:
Usually the backbone is initialized with a pre-trained classifier. So if you're getting results without even doing that stage, then that should give you confidence that your loss layer is working.

Reply:
The code I'm running now uses this pretrained backbone: https://pjreddie.com/media/files/darknet53.conv.74.
I converted it using the Darknet weights visitor and assigned it to the network in this PR.

Reply:
Oh sorry, i misunderstood. Still, there are quite a few layers after that which are randomly initialized. So your loss layer must be working now.